pkg/kv/kvserver/replica_application_result.go

```diff
@@ -129,6 +185,26 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
 				log.Warningf(ctx, "failed to repropose with new lease index: %s", pErr)
 				cmd.response.Err = pErr
 
+				// TODO(tbg): this assertion can mis-fire if we artificially inject
+				// invalid LAIs during proposal, see the explanatory comment above. If,
+				// in the current app batch, we had two identical copies of an entry
+				// (which maps to a local proposal), and the LAI was stale, then both
+				// entries would be local. The first could succeed to repropose, which,
+				// if the propBuf was full, would immediately insert into the proposals
+				// map. Normally, when we then apply the second entry, it would be
+				// superseded and not hit the assertion. But, if we injected a stale LAI
+				// during this last reproposal, we could "accidentally" assign the same
+				// LAI again. The second entry would then try to repropose again, which
+				// is fine, but it could bump into the closed timestamp, get an error,
+				// enter this branch, and then trip the assertion.
+				// We should be able to avoid pathologies such as this one if in
+				// `retrieveLocalProposals` we attached the proposal (and removed from
+				// the map) to the first copy of the current non-superseded entry. This
+				// avoids situations in which multiple entries deal with the same proposal
+				// within the context of an replicaAppBatch.
+				// It might be even better to remove proposals from the map not at the
+				// beginning of application, but at the end: a proposal would leave the
+				// map only once, never to return again.
```

Comment:
This. I think there should be a source-of-truth map that tracks proposals till the end of their lifetime (i.e. inserted once, deleted once). The complexity of the lifetime should be tracked at the level of one entry in this map, we shouldn't be needing to jump in/out of it and have to derive state based on implicit information like "whether it's in the map" / "in the proposal buffer" (this info can probably be abstracted away).

Reply:
Will link this issue here and instead remove the part that talks about possible solutions, since we can just click the link instead.